packages/xforms-engine/src/instance/RankControl.ts

```diff
@@ -0,0 +1,144 @@
+		XFormsXPathElement,
+		EvaluationContext,
+		ValidationContext,
+		ClientReactiveSubmittableValueNode
+{
+	static from(parent: GeneralParentNode, definition: RankDefinition): RankControl {
+		assertRangeNodeDefinition(definition);
+		return new this(parent, definition);
+	}
+
+	private readonly mapOptionsByValue: Accessor<RankItemMap>;
+
+	protected override readonly getInstanceValue: Accessor<string>;
+
+	// XFormsXPathElement
+	override readonly [XPathNodeKindKey] = 'element';
+
+	// InstanceNode
+	protected readonly state: SharedNodeState<RankControlStateSpec>;
+	protected readonly engineState: EngineState<RankControlStateSpec>;
+
+	// RankNode
+	readonly nodeType = 'rank';
+	readonly currentState: CurrentState<RankControlStateSpec>;
+
+	private constructor(parent: GeneralParentNode, definition: RankDefinition<'string'>) {
+		const codec = new BaseItemCollectionCodec(sharedValueCodecs.string);
+		super(parent, definition, codec);
+
+		const valueOptions = createItemCollection(this);
+		const mapOptionsByValue: Accessor<RankItemMap> = this.scope.runTask(() => {
+			return createMemo(() => {
+				return new Map(valueOptions().map((item) => [item.value, item]));
+			});
+		});
+
+		this.mapOptionsByValue = mapOptionsByValue;
+
+		const baseValueState = this.valueState;
+		const [baseGetValue, setValue] = baseValueState;
+		const getValue = this.scope.runTask(() => {
+			return createMemo(() => {
+				const optionsByValue = mapOptionsByValue();
+				return baseGetValue().filter((value) => optionsByValue.has(value));
+			});
+		});
+		const valueState: SimpleAtomicState<readonly string[]> = [getValue, setValue];
+
+		this.getInstanceValue = this.scope.runTask(() => {
+			return createMemo(() => codec.encodeValue(getValue()));
+		});
+
+		const sharedStateOptions = {
+			clientStateFactory: this.engineConfig.stateFactory,
+		};
+
+		const state = createSharedNodeState(
+			this.scope,
+			{
+				reference: this.contextReference,
+				readonly: this.isReadonly,
+				relevant: this.isRelevant,
+				required: this.isRequired,
+				label: createNodeLabel(this, definition as AnyNodeDefinition),
+				hint: createFieldHint(this, definition as AnyNodeDefinition),
+				children: null,
+				valueOptions,
+				value: valueState,
+				instanceValue: this.getInstanceValue,
+			},
+			sharedStateOptions
+		);
+
+		this.state = state;
+		this.engineState = state.engineState;
+		this.currentState = state.currentState;
+	}
+
+	getValueLabel(value: string): string {
+		const valueOption = this.currentState.valueOptions.find(item => item.value === value);
+		return valueOption.label.asString;
+	}
+
+	setValues(valuesInOrder: readonly string[]): Root {
+		const sourceValues: string[] = Array.from(this.mapOptionsByValue().keys());
+		const hasAllValues = !sourceValues.some((sourceValue) => valuesInOrder.includes(sourceValue));
+		if (hasAllValues) {
+			throw new RankFunctionalityError('There are missing options. Rank should have all options.', 'Rank Control');
```

Comment:
As mentioned on inline where the class is defined, I'm not sure about this error's class hierarchy. I think addressing that will help sort out the type error here too.

Reply:
Resolved